Debug-frontend-tests.md

```diff
@@ -0,0 +1,118 @@
+
+**Remember to revert your `fit()` and `fdescribe()` changes before committing!**
+
+Once you've specified which tests you want to run, you should make sure you can still reproduce the bug you are investigating. Some bugs are caused by how tests relate to each other, so in these cases, you won't be able to reproduce the bug by running the failing test in isolation. To see how this might happen, consider the following test code:
+
+```js
+describe('oppia', () => {
+  var test = 2;
+
+  it('should do something', () => {
+    test = 3;
+  });
+
+  it('should do something else', () => {
+    expect(test).toBe(2);
+  });
+});
+```
+
+These tests will fail if `should do something` runs before `should do something else`, but not if the tests run in the reverse order. Our frontend tests run in a non-deterministic order, so these kinds of problems generally appear as flakes. In other words, these tests will sometimes pass but fail other times, even if the underlying code is the same.
+
+## Verbose mode
+
+By default, we suppress log output from the tests to the terminal when running the frontend tests. This keeps the test output clean so it's easy to find which tests failed. However, this output can be useful for debugging. For example, if you add `console.log()` or `console.error()` statements to your tests for debugging, the output from those statements will be suppressed by default. To view this output, pass `--verbose` when you run the frontend tests:
+
+```console
+python -m scripts.run_frontend_tests --verbose
+```
+
+**Only use --verbose when you are running a few tests in isolation. Otherwise, you will be swamped with way too many log messages.**
+
+Note that depending on your situation, `console.log()` or `console.error()` might be preferable:
+
+* `console.log()` statements will cause the linter to fail, so they are a great choice for local debugging. Then if you forget to remove them before pushing your changes, the linter will remind you.
+* `console.error()` statements do not cause a lint failure, so they work well when you are pushing your code with debugging code to a PR to let the tests run in the CI environment.
+
+## Find stack elements
+
+While you can't use the line numbers in a stack trace to find the associated code, you can sometimes use the function or class names. For example, in the example above we could try searching the codebase for `_maybeConvertBody`. The success of this technique depends on how frequently the string you search for occurs in the codebase. If the string occurs a lot, you will have a hard time figuring out which occurrence is the one you are interested in.
+
+## Run tests repeatedly on CI
+
```

Comment:
Does the term "flaky" need to be defined?

Reply:
I just removed it